Executor refactoring

freshtonic opened the pull request:
This is ready for review and merge. Once this is done, I'll rebase the mutations code on top.

freshtonic committed:
Extracted ExecutionContext module from Executor

freshtonic committed:
Refactor: pass and return `ExecutionContext` everywhere

This will allow us to report errors from anywhere in the flow and
accumulate them in the ExecutionContext.

freshtonic commented:
The `Executor` is ugly and needs further refactoring but this is a step that is necessary for correct error reporting. Anything that reports an error has to be able to create a new `ExecutionContext` with a modified `errors` list.

freshtonic committed:
Refactor: prefer a multiple function heads over case

freshtonic committed:
Create `operation_node` type and move it to `Nodes`

freshtonic committed:
Refactor: create Resolvable protocol

freshtonic committed:
Refactor: pattern match source dereferencing

freshtonic committed:
Refactor: use keyword list for optional args

freshtonic committed:
Refactor: assert_execute ignores errors

freshtonic committed:
Assert correct error in mutations test

joshprice commented:
Great work! Since we're changing the primary interface for the lib, we should maintain compatibility functions and deprecate them for the next version.

This will also need the corresponding changes in plug.

This should also resolve #56 and #64 but needs testing. #65 should also pass now so lets pull that into this PR.

freshtonic committed:
Retain old execute function for GraphQL plug compatibility

After new version of plug is released that uses the new execute
function, the old execute function will be deleted.
